README.md
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/randomcoloR)](https://cran.r-project.org/package=randomcoloR)

# randomcoloR

An [R](https://www.r-project.org/) package for generating attractive and distinctive colors.

The `randomColor()` function is ported from [randomColor.js](https://github.com/davidmerfield/randomColor).

Let's quickly get some pretty random colors.
```r
library(igraph)
library(randomcoloR)

k <- 200
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k))
```
![](readme_demo/graph1.png)

We can specify a particular hue, such as red.
```r
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k, hue="red"))
```
![](readme_demo/graph2.png)

We can also get random colors with specific luminosity.
```r
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k, luminosity="light"))
```
![](readme_demo/graph3.png)

We can also ask for a set of optimally distinct colors so that colors in our plot are not too similar.
If we use `ggplot2` to select the color space for our states in the map below, we get many similar colors.
```r
library(dplyr)
library(ggplot2)
library(maps)

states_map <- map_data("state")

ggplot(states_map, aes(x=long, y=lat, group=group, fill=region)) +
  geom_polygon(color="black") +
  guides(fill=FALSE)
```
*Which states are green?*
![](readme_demo/map1.png)

Instead, let's find the most distinctive set of colors for all states.
```r
ggplot(states_map, aes(x=long, y=lat, group=group, fill=region)) +
  geom_polygon(color="black") +
  scale_fill_manual(values=distinctColorPalette(length(unique(states_map$region)))) +
  guides(fill=FALSE)
```
*Now, which states are green?*
![](readme_demo/map2.png)

When using `ggplot2`, we can specify a custom color palette.
```r
ggplot(mtcars, aes(x=disp, y=mpg, col=as.factor(gear))) +
  geom_point(size=5) +
  scale_colour_manual(values=randomColor(length(unique(mtcars$gear)), luminosity="light")) +
  theme_bw()
```
![](readme_demo/mtcars_custom_palette.png)

You can use the default color space.

```r
set.seed(8675309)

scales::show_col(distinctColorPalette(12), labels=FALSE)
```

![](readme_demo/default_cs_12.png)

Or an alternate color space.

```r
scales::show_col(distinctColorPalette(12, altCol=TRUE), labels=FALSE)
```

![](readme_demo/alt_cs_12.png)

And you can preprocess the color space with t-SNE for improved color separation
in some circumstances

```r
scales::show_col(distinctColorPalette(12, altCol=TRUE, runTsne=TRUE), labels=FALSE)
```

![](readme_demo/alt_cs_tsne_12.png)

## Installation from Github

To install this package from Github via the R console, type:
```r
devtools::install_git("https://github.com/ronammar/randomcoloR")
```
It's also on [CRAN](https://cran.r-project.org/web/packages/randomcoloR/):
```r
install.packages("randomcoloR")
```


NEWS
randomcoloR 1.1.0
----------------------------------------------------------------

NEW FEATURES

* Alternate color space for distinct colors

* t-SNE preprocessing of color space for improved color separation

MINOR IMPROVEMENTS

* distinctColorPalette() no longer outputs a named vector, improving compatibility
  with ggplot2 (where color names are meaningful)

* Shrinks original color space for improved performance (especially with t-SNE
  and k-medoids)


R/randomcolor.R
# Author: Ron Ammar
# Description: Get pretty random colors in R. Wraps randomColor.js.

#' Get a pretty random color.
#'
#' @param count number of colors (>= 1)
#' @param hue The hue of the colors to be selected.
#' @param luminosity The luminosity of the colors to be selected.
#' @return A character vector of random color hexadecimal codes.
#' @examples
#' randomColor()
#'
#' randomColor(hue="pink")
#'
#' randomColor(10, luminosity="light")
#' @export
#' @import stringr
randomColor <- function(count=1,
                        hue=c(" ", "random", "red", "orange", "yellow", "green", "blue", "purple", "pink", "monochrome"),
                        luminosity=c(" ", "random", "light", "bright", "dark")
) {
  hue <- match.arg(hue)
  luminosity <- match.arg(luminosity)

  str_split(ct$eval(paste0("randomColor({hue:'", hue, "',",
                           "luminosity:'", luminosity,"',",
                           "count:", count,"})")
  ), ",")[[1]]
}


#' Generate palettes of optimally distinct colors.
#' @description Inspired by the the theory from http://tools.medialab.sciences-po.fr/iwanthue/theory.php
#' For more info, also see https://en.wikipedia.org/wiki/Lab_color_space
#'
#' @param k number of colors (>= 1). May be ineffective for k > 40.
#' @param altCol Use an alternate color space
#' @param runTsne Preprocess color space with t-SNE to obtain distinct colors. Reduces performance.
#' @return A character vector of k optimally distinct colors in hexadecimal codes.
#' @export
#' @import cluster Rtsne
distinctColorPalette <-function(k=1, altCol=FALSE, runTsne=FALSE) {
  currentColorSpace <- ourColorSpace@coords
  if (altCol) {
    currentColorSpace <- alternateColorSpace
  }

  if (runTsne) {
    # Run 2D t-SNE before clustering
    tsne <- Rtsne(currentColorSpace, perplexity=50, check_duplicates=FALSE, pca=FALSE, max_iter=500)
    pamx <- pam(tsne$Y, k)  # k-medoids
    if (altCol) {
      colors <- rgb(currentColorSpace[pamx$id.med, ], maxColorValue=255)
    } else {
      colors <- hex(LAB(currentColorSpace[pamx$id.med, ]))
    }
  } else {
    # Set iter.max to 20 to avoid convergence warnings.
    km <- kmeans(currentColorSpace, k, iter.max=20)
    if (altCol) {
      colors <- rgb(round(km$centers), maxColorValue=255)
    } else {
      colors <- unname(hex(LAB(km$centers)))
    }
  }

  return(colors)
}


readme_demo/demo.R
# These are examples of randomColors in action

library(Cairo)
library(igraph)
library(randomcoloR)

k <- 200

CairoPNG("demo/graph1.png")
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k))
dev.off()

CairoPNG("demo/graph2.png")
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k, hue="red"))
dev.off()

CairoPNG("demo/graph3.png")
plot(erdos.renyi.game(k, 0.1), vertex.label=NA,
     edge.lty="blank", vertex.color=randomColor(k, luminosity="light"))
dev.off()


library(dplyr)
library(ggplot2)
library(maps)

states_map <- map_data("state")

# Using the default ggplot scheme, many similar colors for states
ggplot(states_map, aes(x=long, y=lat, group=group)) +
  geom_polygon(aes(fill=region), color="black") +
  guides(fill=FALSE)
ggsave("demo/map1.png", device="png", width=7, height=4)

# Instead, let's find the most distinctive set of colors for all states
# Which state is pink below?
s <- unique(states_map$region)
df <- data.frame(region=s, newColor=distinctColorPalette(length(s)),
                 stringsAsFactors=FALSE)
states_map <- left_join(states_map, df, by="region")

# Now, which state is pink below? The colors are more optimally distinct now.
ggplot(states_map, aes(x=long, y=lat, group=group)) +
  geom_polygon(fill=states_map$newColor, color="black")
ggsave("demo/map2.png", device="png", width=7, height=4)
